### ISSUE TITLE
Invalid token on payment

### ISSUE THREAD

**Issue by jonnoconn** [dualcube/moodle-enrol_stripepayment#138]
Hi,
I'm getting an error after payment:

```
{
    "exception": "moodle_exception",
    "errorcode": "invalidtoken",
    "message": "Invalid token - token not found"
}
```

That appears directly after the payment window. Stripe does take the payment (I've checked with live accounts).

Some settings:
- I have enabled web-services as well as REST APIs in /admin/settings.php?section=webserviceprotocols . 
- I have created a token that is the service Moodle_enrol_stripe_payment as well as moodle_enrol_stripepaymentpro and Moodle mobile web service . None of them work!
- I am running Moodle 4.3
- If I use a 100% coupon, everything works correctly and the person is enrolled, no error messages. 

Ideas suggestions? Most of the documentation I've seen from other issues related to not enabling web services.

**Comment by moumitahalder** [comment]
@jonnoconn, I'm sorry to hear that you're experiencing this issue.

Based on the error message, it appears that the token you added during the configuration of Stripe may no longer be valid. Could you please refer to our documentation to ensure that the token is added correctly : 
 https://dualcube.com/docs/stripe-payment-with-sca-and-coupon/#3-toc-title

Please do keep us posted.

**Comment by jonnoconn** [comment]
Firstly, thank you for your patience. 

I've been testing different variations of tokens. I've even deleted all payment options and started from scratch. I am just trying to get stripe-payment-with-sca-and-coupon working now. Once that is going, I'm sure the StripePaymentPro should be fine. 
Some quick notes:
- The system will accept payment and enrol the person f I do an `enrol on payment` using PayPal or stripe (which excludes this plugin). This makes me believe my config/setup is fine in stripes. 
- If I add a 100% discount, the transaction is processed successfully. 50% discount coupons fail as per normal payment. 

Some more on my setup:
- Rest API is enabled. SOAP is disabled.
- The `moodle_enrol_stripepayment` token has no expiry date. 
- The token is owned by me, the admin. 
- I have 1 other token: A `Moodle mobile web service` token
- The payment is processed by Stripe. i.e. within stripe, I can see the "PaymentIntent status: succeeded" on the transaction. In live tests, money leaves the account. 
- Stripe has a flag that says "`Pending webhook response. Several webhooks that are subscribed to the event haven't successfully responded yet`"
- In the `Stripe-payment-with-sca-and-coupon` setup, it says I need to add the [Client ID], but there is no where for it to go. 

Something else I've noticed going through the stripe logs:
When I deep dive into Stripe's logs, I see the specific call that fails to generate a response. Specifically, it says in the logs that the following web hook did not respond (error 404): `https://learning.oya.team/webservice/rest/server.php?<my moodle_enrol_stripepayment token name>&wsfunction=moodle_stripepayment_success_stripe_url&moodlewsrestformat=json&session_id={CHECKOUT_SESSION_ID}&user_id=13&couponid=&instance_id=167`

Any ideas?

**Comment by moumitahalder** [comment]
@jonnoconn as you can see you we are able to successfully place 10% coupon code : http://tinyurl.com/ytcb72os'

User is able to access the course also : http://tinyurl.com/2x7ebol3

Now as you are facing this issue, please do contact use [here ](https://dualcube.com/forums/support-forum), so we can assist you with this.

**Comment by jonnoconn** [comment]
ok, solved the problem. It had to do with tokens. I am not sure why I was seeing this error. 

**It FAILED when:** 
- I created a token using my own admin account
- I only saw the token name, nothing else, so that is what I used. 
![Main-admin screen](https://github.com/dualcube/moodle-enrol_stripepayment/assets/42315771/bdfc5064-0c1e-41f4-aa07-16de08b5cd48)

**It WORKED when:**
- I created a specific admin account, just for managing tokens. 
- When I created a token using this account, it displayed a new section (unseen on my main admin) that had token name + token code. 
- Using this token code, everything worked fine.
![Token-admin screen](https://github.com/dualcube/moodle-enrol_stripepayment/assets/42315771/ccb66767-e6e4-412e-8f4c-cac027e6b97a)

**Advice for future users**
If anyone encounters this problem in the future, create a new admin account specifically for token creation, and create the tokens using that account.
